### ISSUE TITLE
Add( update:true) list updating confusion

### ISSUE THREAD

**Issue by AndyDentFree** [realm/realm-dotnet#1147]
We may have a design problem or need to add another variant for this case.

A common scenario for using the `update:true` flag is people importing data. A user reports (note he is using a copy of the 0.82.0 NuGet I gave him which already has the recent IList fixes in #1040):

Basically I have an object called Conversation that contains an IList of objects called Message, essentially the structure of a messaging app like Whatsapp.

Now every time I fetch the `Conversation` 's from the web-api (our server), which fetches any new or updated `Conversation` 's since the last checked date and any new `Message` 's. I call `Realm.Manage(conversations,true);`.

As you would expect this updates the relevant `Conversation` but when it manages the new `Conversation` object. The only `Message` objects it contains are the new ones from the server, in its IList property.

So it looks like instead of executing the same update behavior on the child objects (add the new ones, update the existing ones). It is just overwriting the IList with the the list of new `Message` 's. Shouldn't it be adding to or updating them?

The only solution I can think of currently is retrieve the existing `Conversation` from realm and copy across the list from that in to the new `Conversation` list then manage all of that. Either that or redesign the whole thing.

Here's the Conversation object for reference

```
   [Preserve(AllMembers = true)]
    public class Conversation : RealmObject
    {
        [PrimaryKey]
        public string Id { get; set; }
        public string Subject { get; set; }
        public string Advisor { get; set; }
        public string Client { get; set; }
        public DateTimeOffset RelevanceDateUtc { get; set; }
        public bool IsRead { get; set; }
        public IList<Message> Messages { get; }
   }
```

**Comment by nirinchev** [comment]
Is the user building from source? As far as I remember, the current nuget has a bug that actually merges `IList` properties (so should work as the user desires). This, however, was fixed in #1040, so with our next version lists will not be merged, but overwritten instead (i.e. what the user describes).

Currently `Add(item, update: true)` is designed to work as `PUT` - replacing the whole object as opposed to doing partial updates. In that aspect, the current behavior (replacing the content of the list instead of merging it) is correct.

That being said, it doesn't help the user much :) So we have two options - add an optional `mergeChildLists` parameter to our `Add` (don't like it much) or advise them to redesign their app to use backlinks.

So essentially, the model would be:
```csharp
public class Conversation : RealmObject
{
    /* other properties */
    [Backlink(nameof(Message.Conversation))]
    public IQueryable<Message> Messages { get; }
}
```

and for their `Message`:
```csharp
public class Message
{
    public Conversation Converation { get; set; }
}
```

That means they would have to remodel their json to disjoin the message and conversation objects:

```json
{
    "conversation": {
        "id": "123",
        ...
    },
    "messages": [
        {
            "id": "456",
            ...
        }
    ]
}
```

Finally, he would then need to do:
```csharp
var conversation = realm.Add(jsonResponse.Conversation, update: true);

foreach (var message in jsonResponse.Messages)
{
    message.Conversation = conversation;
    realm.Add(message);
}
```

where `jsonResponse` is some wrapper object:
```csharp
public class ConversationMessageWrapper // could figure out a better name :D
{
    public Conversation Conversation { get; set; }
    
    public IEnumerable<Message> Messages { get; set; }
}
```

**Comment by AndyDentFree** [comment]
As I mentioned in the report (as an edit, so you may not have noticed), he has the #1040 fix already.

I'd like to check our current list-replacement behaviour against the other SDKs to be sure it's correct.

Because we don't have any schema flag of _ownership semantics_  it is arguable that our current Add/update behaviour may be wrong. If chlid objects were _owned_ by the parent, we should be deleting them when we replace an entire list, if they are not in the incoming set.

In summary, I think there's a few more dimensions to this and will think about it a bit more but at least your backlink suggestion gives them another model to try.

**Comment by nirinchev** [comment]
I am fairly certain our current behavior is aligned with other SDKs (here's [Cocoa](https://github.com/realm/realm-cocoa/blob/dfbec5b9f3c82280bf4e4e3dda5e4ca5b7d72945/Realm/RLMAccessor.mm#L7510)).

As mentioned, replacing the list is the correct behavior for the current semantics of update, because, after an update, the state of the object should be as provided before the update.

When we add support for cascading deletes, we could use that as an indication for deleting child objects, but until then, it's not safe to do it.
